### PR TITLE
implemented DAO (CRUD), created connection to DB.

### PR DESCRIPTION
**Pull request by Daniel1004K**

### CODE REVIEW

File: src/main/java/mate/jdbc/Main.java [hunk 1]
```diff
@@ -1,7 +1,31 @@
...
+import mate.jdbc.dao.ManufacturerDao;
+import mate.jdbc.dao.ManufacturerDaoImpl;
+import mate.jdbc.lib.Injector;
+import mate.jdbc.model.Manufacturer;
+
 public class Main {
+    private static final Injector injector = Injector.getInstance("mate.jdbc");
+
     public static void main(String[] args) {
+        ManufacturerDao manufacturerDao = new ManufacturerDaoImpl();
```
Comment (deniskichan) [inline src/main/java/mate/jdbc/Main.java:13]: use injector for creating instances

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+import java.util.Optional;
+import mate.jdbc.lib.Dao;
+import mate.jdbc.model.Manufacturer;
+import mate.jdbc.util.ConnectionUtil;
+
+@Dao
+public class ManufacturerDaoImpl implements ManufacturerDao {
+    public static final String COLUMN_NAME = "name";
+    public static final String COLUMN_COUNTRY = "country";
+    public static final String COLUMN_ID = "id";
```
Comment (deniskichan) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:19]: these constants are redundant here

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    }
+
+    @Override
+    public boolean delete(Long id) {
+        String request = "UPDATE manufacturers SET is_deleted = TRUE WHERE id = ?";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement query = connection.prepareStatement(request)) {
+            query.setObject(1, id);
+            int updateRows = query.executeUpdate();
+            return updateRows > 0;
```
Comment (deniskichan) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:95]:
```suggestion
            return query.executeUpdate() > 0;
```

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+import java.util.Optional;
+import mate.jdbc.lib.Dao;
+import mate.jdbc.model.Manufacturer;
+import mate.jdbc.util.ConnectionUtil;
+
+@Dao
+public class ManufacturerDaoImpl implements ManufacturerDao {
+    public static final String COLUMN_NAME = "name";
+    public static final String COLUMN_COUNTRY = "country";
+    public static final String COLUMN_ID = "id";
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:19]: better make them private

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        while (resultSet.next()) {
+            String name = resultSet.getString(COLUMN_NAME);
+            String country = resultSet.getString(COLUMN_COUNTRY);
+            Long id = resultSet.getObject(COLUMN_ID, Long.class);
+            manufacturerList.add(new Manufacturer(id, name, country));
+        }
+        return manufacturerList;
+    }
+
+    private void setKey(Manufacturer manufacturer, ResultSet generatedKeys) throws SQLException {
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:111]: цей метод використовуэться лише один раз в одному методі, не бачу особливої користі виносити логіку в окремий метод

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement preparedStatement = connection.prepareStatement(request,
+                        Statement.RETURN_GENERATED_KEYS)) {
+            preparedStatement.setString(1, manufacturer.getName());
+            preparedStatement.setString(2, manufacturer.getCountry());
+            preparedStatement.executeUpdate();
+            ResultSet generatedKeys = preparedStatement.getGeneratedKeys();
+            setKey(manufacturer, generatedKeys);
+        } catch (SQLException e) {
+            throw new RuntimeException("Can't operate with DB", e);
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:33]: image
https://mate-academy.github.io/jv-program-common-mistakes/java-JDBC/jdbc-intro/JDBC-intro_checklist.html

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        return manufacturer;
+    }
+
+    @Override
+    public Optional<Manufacturer> get(Long id) {
+        String request = "SELECT * FROM manufacturers where id= ? AND is_deleted = FALSE;";
+        try (Connection connection = ConnectionUtil.getConnection();
+                PreparedStatement preparedStatement = connection.prepareStatement(request)) {
+            preparedStatement.setObject(1, id);
+            ResultSet generatedKeys = preparedStatement.executeQuery();
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:44]:
```suggestion
            ResultSet resultSet = preparedStatement.executeQuery();
```

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+            if (generatedKeys.next()) {
+                String name = generatedKeys.getString(COLUMN_NAME);
+                String country = generatedKeys.getString(COLUMN_COUNTRY);
+                return Optional.of(new Manufacturer(id, name, country));
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:48]: image
https://mate-academy.github.io/jv-program-common-mistakes/java-JDBC/jdbc-intro/JDBC-intro_checklist.html
we can use this private method in the get() and getAll() methods

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    private List<Manufacturer> getManufacturers(ResultSet resultSet) throws SQLException {
+        List<Manufacturer> manufacturerList = new ArrayList<>();
+        while (resultSet.next()) {
+            String name = resultSet.getString(COLUMN_NAME);
+            String country = resultSet.getString(COLUMN_COUNTRY);
+            Long id = resultSet.getObject(COLUMN_ID, Long.class);
+            manufacturerList.add(new Manufacturer(id, name, country));
+        }
+        return manufacturerList;
+    }
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:109]:
```suggestion
    private Manufacturer getManufacturer(ResultSet resultSet) throws SQLException {
            String name = resultSet.getString(COLUMN_NAME);
            String country = resultSet.getString(COLUMN_COUNTRY);
            Long id = resultSet.getObject(COLUMN_ID, Long.class);
            return new Manufacturer(id, name, country);
    }
```
and then we can use this method in the get() and getAll() methods

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    public List<Manufacturer> getAll() {
+        List<Manufacturer> manufacturers;
+        String request = "SELECT * FROM manufacturers WHERE is_deleted = FALSE";
+        try (Connection connection = ConnectionUtil.getConnection();
+                Statement statement = connection.createStatement()) {
+            ResultSet resultSet = statement
+                    .executeQuery(request);
+            manufacturers = getManufacturers(resultSet);
+        } catch (SQLException e) {
+            throw new RuntimeException("Can't operate with DB", e);
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:66]:
```suggestion
            throw new RuntimeException("Can't get all manufacturers from DB", e);
```

File: src/main/java/mate/jdbc/model/Manufacturer.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+public class Manufacturer {
+    private Long id;
+    private String name;
+    private String country;
+
+    public Manufacturer(Long id, String name, String country) {
+        this.id = id;
+        this.name = name;
+        this.country = country;
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/model/Manufacturer.java:13]:
```suggestion
        this(name, country);
        this.id = id;
```

File: src/main/java/mate/jdbc/util/ConnectionUtil.java [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.Properties;
+
+public class ConnectionUtil {
+    public static final String LOGIN = "root";
+    public static final String PASSWORD = "password";
+    public static final String JDBC_DRIVER = "com.mysql.cj.jdbc.Driver";
+    public static final String MANUFACTURER_DB = "jdbc:mysql://localhost:3306/manufacturer_db";
+    public static final String LOGIN_PROPERTY = "user";
+    public static final String PASSWORD_PROPERTY = "password";
+    public static final String EXCEPTION_MESSAGE = "Can't connect to DB";
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/util/ConnectionUtil.java:15]: this is redundant to move exception message to the constant

**Review by Sofasmile — CHANGES_REQUESTED** [review]

File: src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java [hunk 13]
```diff
@@ -28,9 +29,12 @@ public Manufacturer create(Manufacturer manufacturer) {
...
             preparedStatement.executeUpdate();
             ResultSet generatedKeys = preparedStatement.getGeneratedKeys();
-            setKey(manufacturer, generatedKeys);
+            if (generatedKeys.next()) {
+                Long id = generatedKeys.getObject(1, Long.class);
+                manufacturer.setId(id);
+            }
         } catch (SQLException e) {
-            throw new RuntimeException("Can't operate with DB", e);
+            throw new DataProcessingException("Can't operate with DB", e);
```
Comment (Sofasmile) [inline src/main/java/mate/jdbc/dao/ManufacturerDaoImpl.java:37]:
```suggestion
            throw new DataProcessingException("Can't insert to DB " + manufacturer, e);
```
all messages should be informative

**Review by Sofasmile — APPROVED** [review]

**Pull request closed without merging**